atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -41,25 +41,92 @@
  */
 public interface Transaction {
 
+    /**
+     * Returns a mapping of rows to {@link RowResult}s within {@code tableRef} for the specified {@code rows}, loading
+     * columns according to the provided {@link ColumnSelection}.
+     *
+     * The returned {@link SortedMap} is sorted on the byte order of row keys; the ordering of the input parameter
+     * {@code rows} is irrelevant.
+     *
+     * If there are rows with no cells matching the provided {@link ColumnSelection}, they will not be present in the
+     * {@link Map#keySet()} of the output map at all.
+     *
+     * @param tableRef table to load rows from
+     * @param rows rows to be loaded
+     * @param columnSelection columns to load from the given rows
+     * @return a mapping of rows to the columns matching the provided column selection
+     */
     @Idempotent
     SortedMap<byte[], RowResult<byte[]>> getRows(
             TableReference tableRef,
             Iterable<byte[]> rows,
             ColumnSelection columnSelection);
 
+    /**
+     * Returns a mapping of rows to {@link BatchingVisitable}s (which may be thought of as iterators over columns)
+     * within {@code tableRef} for the specified {@code rows}, where the columns fall within the provided
+     * {@link BatchColumnRangeSelection}. The single provided {@link BatchColumnRangeSelection} applies to all of the
+     * rows.
+     *
+     * The returned {@link BatchingVisitable}s are guaranteed to return cells matching the predicate, sorted on
+     * (ascending) byte ordering.
+     *
+     * It is guaranteed that the {@link Map#keySet()} of the returned map has the same elements as {@code rows},
+     * even if there are rows where no columns match the predicate.
```

Comment:
@jkozlowski notes: does same elements here mean the same *objects*, or just arrays that are equal in terms of `Arrays#equals()`? I intended the former, historically the behaviour of getRowsColumnRange is as follows:

```
    @Test
    public void identity() {
        txMgr.runTaskWithRetry((tx) -> {
            put(tx, "tom", "aaa", "42");
            return null;
        });

        txMgr.runTaskWithRetry((tx) -> {
            byte[] b1 = PtBytes.toBytes("tom");
            Map<byte[], BatchingVisitable<Map.Entry<Cell, byte[]>>> iterators = tx.getRowsColumnRange(TEST_TABLE,
                    ImmutableList.of(b1),
                    BatchColumnRangeSelection.create(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY, 10));

            byte[] b1v2 = PtBytes.toBytes("tom");
            assertThat(iterators.get(b1)).isNotNull(); // Passes
            assertThat(iterators.get(b1v2)).isNotNull(); // Fails
            return null;
        });
    }
```

We should probably be stricter/more explicit that the identity of the rows *must* be equal, will incorporate this.

Reply:
To be clear, unless people feel strongly about this I have no intention of codifying the failure of the second assertion above - but will say that the map must be accessed by the original rows passed in, and using other rows (even if equal in terms of byte arrays to an original row) is undefined behaviour.

Reply:
Yes, we should make it clear that pulling things out by the original byte arrays is fine.

Reply:
Agreed. My point is also that pulling things out by means other than the byte arrays is a bad idea, though I don't want to guarantee that it _will_ fail.